Adding classification tags options to `ica_reclassify`

handwerkerd opened the pull request:
Closes #1182.
This is in response to #1182 and discussion in #1184.

Changes proposed in this pull request:

- `ica_reclassify` gets two new command line options `--tagacc` and `--tagrej`. Following the confection for `--manacc`, these are `tag_accept` and `tag_reject` in the workflow
- When `ica_reclassify` is run these tags will be added to all listed components, whether or not the classification changes. That means one could add a tag like "AROMA reject" and it would be added to the existing tags whether or not a component was already classified as rejected.
  - This was always the desired functionality within the API, but I noticed & corrected an edge case where this didn't work & now it is accessible via the CLI
- The inputted tags can be either a single tag or a comma delimited string (i.e. `--tagrej "AROMA reject reason 1, AROMA reject reason 2"`). I think this makes the full functionality @tsalo wants possible, but you'd need to run & write out new outputs for every different combination of new tags. This is inefficient.
- tests added for all new functionality and some existing tests were improved to confirm details of outputs while I was editing them.

handwerkerd commented:
Same CircleCI testing issue (see this comment: https://github.com/ME-ICA/tedana/pull/1199#issuecomment-2749403194 )

Also, codecov has the missing head commit message again (#1192) https://app.codecov.io/gh/ME-ICA/tedana/pull/1200

handwerkerd commented:
@tsalo This is ready for review. I'm not sure if you're going to use this feature, but I figured you'd be a good reviewer. If you don't have time, I can swap you out for someone else.

handwerkerd commented:
@eurunuela after #1202, Codecov is properly integrated and I noticed a few new lines that tests missed. The small, but nice pleasure of getting 100% testing coverage. 😄

tsalo commented on tedana/selection/selection_utils.py:
```diff
@@ -253,6 +253,27 @@ def comptable_classification_changer(
             selector.component_table_.loc[changeidx, "classification"] = classify_if
+
+        if tag_if is not None:  # only run if a tag is provided
+            # if tag_if has commas, divide into multiple tags
+            if "," in tag_if:
+                multi_tags = tag_if.split(",")
+                tag_if = set([s.strip() for s in multi_tags])
+            else:
+                tag_if = {tag_if}
```
Is this necessary? If the input is a comma-delimited string and the field in the component table is a comma-delimited string, do you need to modify it?

handwerkerd replied:
This makes sure the same tag isn't repeated. That is, if the existing table has `tag1,tag2` and the new tags are `tag2,tag3` the result will be `tag1,tag2,tag3` without a repetitionof `tag2`

tsalo replied:
Ah okay, thanks. I didn't realize this happened after the old and new tags were concatenated.

tsalo commented on tedana/selection/selection_utils.py:
```diff
@@ -253,6 +253,27 @@ def comptable_classification_changer(
+                tag_if = set([s.strip() for s in multi_tags])
+            else:
+                tag_if = {tag_if}
+            for idx in changeidx:
```
```suggestion

            for idx in changeidx:
```

tsalo commented on tedana/tests/test_integration.py:
```diff
@@ -188,19 +188,27 @@ def test_integration_four_echo(skip_integration):
-    fn = resource_filename("tedana", "tests/data/fiu_four_echo_outputs.txt")
-
-    check_integration_outputs(fn, out_dir)
+    component_table = pd.read_csv(op.join(out_dir_manual, "desc-tedana_metrics.tsv"), sep="\t")
```
```suggestion
    component_table = pd.read_table(op.join(out_dir_manual, "desc-tedana_metrics.tsv"))
```

tsalo commented on tedana/workflows/ica_reclassify.py:
```diff
@@ -232,16 +258,72 @@ def _parse_manual_list(manual_list):
+            elif "," in x:
+                raise ValueError(
+                    "_parse_manual_list_str includes a comma in a list of multiple strings. "
+                    "Input can include a comma deliminated string, but not multiple strings. "
```
```suggestion
                    "Input can include a comma delimited string, but not multiple strings. "
```

tsalo commented on tedana/workflows/ica_reclassify.py:
```diff
@@ -232,16 +258,72 @@ def _parse_manual_list(manual_list):
     return manual_nums
 
 
+def _parse_manual_list_str(manual_list):
```
It might be good to write a unit test for this function. It would be simpler to just test this function directly than to incorporate the test into the integration tests for ica_reclassify.

tsalo reviewed (commented):
I have a couple of minor suggestions/questions, but overall this makes sense to me.

handwerkerd commented:
I think I've addressed all of @tsalo's feedback. To simplify the tests, I moved two functions in `ica_reclassify.py` into `workflows/parser_utils.py` and then added unit tests into `test_workflows_parser_utils.py`. This also let me remove a few integration tests for `ica_reclassify` that were just being used to test failures of the above functions.

FWIW, in setting this all up, I noticed that the docstring for `parse_manual_list_int`said it would identify delimiters that were commas, spaces, or tabs, but the code only used commas for delimiters. Since this hasn't caused any problems, I just updated the docstring to only mention commas.

tsalo reviewed (approved):
LGTM. I'm happy to have this close #1182.
